api/src/main/java/org/eclipse/microprofile/reactive/messaging/connector/IncomingConnectorFactory.java

```diff
@@ -0,0 +1,119 @@
+ * configuration, the reactive messaging implementation would need to find the {@link IncomingConnectorFactory} returning
+ * the {@code i.e.m.reactive.messaging.impl.kafka.Kafka} class as the result of its {@link #type()} method to create the
+ * {@code my-channel} channel. Note that if the connector cannot be found, the deployment must be failed with a
+ * {@link DeploymentException}.
+ * <p>
+ * The {@link #getPublisherBuilder(Config)} is called for every channel that needs to be created. The {@link Config} object
+ * passed to the method contains a subset of the global configuration, and with the prefixes removed. So for the previous
+ * configuration, it would be:
+ * <pre>
+ * bootstrap-services = localhost:9092
+ * topic = my-topic
+ * </pre>
+ * <p>
+ * In this example, if {@code topic} was missing as a configuration property, the Kafka connector would be at liberty to
+ * default to the stream name indicated in the annotation as the Kafka topic. Such connector specific behaviours are
+ * outside the scope of this specification.
+ * <p>
+ * So the connector implementation can retrieve the value with {@link Config#getValue(String, Class)} and
+ * {@link Config#getOptionalValue(String, Class)}.
+ * <p>
+ * If the configuration is invalid, the {@link #getPublisherBuilder(Config)} method must throw an
+ * {@link IllegalArgumentException}, caught by the reactive messaging implementation and failing the deployment by
+ * throwing a {@link DeploymentException} wrapping the exception.
+ * <p>
+ * Note that a Reactive Messaging implementation must support the configuration format described here. Implementations
+ * are free to provide additional support for other approaches.
+ */
+public interface IncomingConnectorFactory {
+
+    /**
+     * Gets the {@link MessagingProvider} class associated with this {@link IncomingConnectorFactory}.
+     * Note that the {@link MessagingProvider} is a user-facing interface used in the configuration.
+     *
+     * @return the {@link MessagingProvider} associated with this {@link IncomingConnectorFactory}. Must not be
+     * {@code null}. Returning {@code null} will cause a deployment failure.
+     */
+    Class<? extends MessagingProvider> type();
+
+    /**
+     * Creates a <em>channel</em> for the given configuration. The given configuration {@code type} attribute matches the
+     * {@link #type()}.
+     * <p>
+     * Note that the connection to the <em>transport</em> or <em>broker</em> is generally postponed until the
+     * subscription occurs.
+     *
+     * @param config the configuration, must not be {@code null}
+     * @return the created {@link PublisherBuilder}, will not be {@code null}.
+     * @throws IllegalArgumentException if the configuration is invalid.
+     * @throws NoSuchElementException   if the configuration does not contain an expected attribute.
+     */
+    PublisherBuilder<? extends Message> getPublisherBuilder(Config config);
```

Comment:
Why should we pass in a config? The config object can be obtained by impl without any problem though.

Reply:
also can you move the spi to spi package?

Reply:
Done in my last push.